### PR TITLE
[POC] Use otf2-template to generate files

### PR DESCRIPTION
**Pull request by Flamefire**
**DO NOT MERGE** Proof-of-concept only!

There is a lot of repetition in the source files and a lot of manual specifying of types. One big problem e.g. is the trait `reference_type` which uses hard-coded types instead of the OTF2 ones. Those could become out of sync w/o anyone noticing.

This is a preview (based on #15) which aims to solve this by using the `otf2-template` script which is installed with OTF2. Compare the size of `reference.inc.tmpl.hpp` with the current `reference.hpp` to clearly see the benefits: A lot less maintenance work and up2date files.

The idea is:
- Define the traits in static include files (only the struct itself)
- Have the specializations auto-generated by the template parser
- Format those with clang-format (BTW: I'd vouch for putting empty class braces on the same line as the class, makes the traits more readable)
- Include that file in the static include (most of the includes in the generated files are for easier validation with e.g. IDEs)
- Install both together (not yet done)

Final note: We were repeatably talking about having official C++ bindings with OTF2. Auto-Generating the files is another step into that direction.

### CODE REVIEW

**Comment by bmario** [comment]
Where shall I begin... 

From my experience in developing the python OTF2 bindings and the C++ wrapper, I am not convinced that the generation is worth it at this point for the C++ wrapper. I'm pretty sure porting the code to the template engine will be a lot more work than it could save even in the long run. I think it will be easier to just add the missing records when they appear in OTF2 by hand. Sure, the argument is that new records will be autogenerated, but that doesn't even work with the Python bindings, why should it with C++ then. Some, but probably not all issues:

- the template engine already proofed itself insufficiently, when we used it for Python. C++ will be even harder, as it is statically typed, uses a lot of templates and in general has more quirks. I mean, yes you can generate the references.hpp, which is a fairly trivial file, try the same with definitions.
- at this point in the lifetime of otf2xx, switching to the template engine is only about deleting existing code, which works well enough in several programs. It doesn't help in developing anymore.
- it is a major rewrite of the whole codebase, which will introduce new bugs

These issues aren't impossible to solve, but I think it's not worth the effort in the end.

And finally:
> We were repeatably talking about having official C++ bindings with OTF2. Auto-Generating the files is another step into that direction.

Nobody talked about that with me, ever. If there are directions, we should have a broad discussion in the next tools meeting.

**Comment by Flamefire** [comment]
Summary of my arguments from the last call:

- There is a lot of repetition in the code, using the template engine reduces this to the bare minimum
- No need to use the engine for everything, just for the core parts (like traits) where the C++ bindings work with the C interface
- (Addition to above): Updating the OTF2 version will be less error-prone. E.g. for some references the type was changed which would go unnoticed by the current version

>at this point in the lifetime of otf2xx, switching to the template engine is only about deleting existing code, which works well enough in several programs. It doesn't help in developing anymore.

- Yes and no. I'd mainly target highly repetitive code. This will decrease maintenance burden in the future.   
And E.g. currently the trait I changed here would need to be updated to use the `OTF2_*` types  to be correct. The work amount for that is very similar and the template version is easier to verify (no C&P errors possible)

Yes, a meeting is required to discuss the preferred direction to go.

**Comment by bmario** [comment]
There are a few pitfalls. I'd like to see it done from beginning to end for one file, a few steps are missing.

- Remove the existing code and include the generated file properly
- workaround for missing definitions (rma_window, call_site, call_path come to my mind)
- using clang-format on the template file breaks horribly; my editor is configured to clang-format every C++ file on save. I think adding `// clang-format off` and `// clang-format on` is justifiable.
- add the file to install target

File: templates/include/otf2xx/traits/reference.inc.tmpl.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    // Type traits for Definitions
+    @otf2 for def in defs|referable_defs:
+    @otf2  if def.lower == 'group':
+    template <typename Def, otf2::common::group_type Type>
+    struct reference_type<definition::@@def.lower@@<Def, Type>> : identity<@@def.self_attr.type@@>{};
+    @otf2  else:
+    template <>
+    struct reference_type<definition::@@def.lower@@> : identity<@@def.self_attr.type@@>{};
+    @otf2  endif
+    @otf2 endfor
```
Comment (bmario) [inline templates/include/otf2xx/traits/reference.inc.tmpl.hpp]: Adding a newline here improves the readability of the generated code a lot

**Comment by Flamefire** [comment]
Will do.

> using clang-format on the template file breaks horribly; my editor is configured to clang-format every C++ file on save. I think adding // clang-format off and // clang-format on is justifiable.

Another idea: Call the file `reference.inc.hpp.tmpl` instead. This would solve the format problem, and make the glob easier.  
Downside: We might want to mix generated code and not generated code in which case we can't format neither, but it is a template anyway.

Wait: If I'd add the `clangformat off` tags the resulting source would also not be formatted which is bad. So I'd got with `*.tmpl` as the extension and make it a rule not to mix generated and written code as far as possible.

**Comment by tilsche** [comment]
You have to consider the inevitable cost of introducing the template generation into this project. This includes the additional CMake code that needs to be maintained, the template source files, als the project now requires `clang-format` to be *built*.

I don't see how generating the small subset of regular code that requires no exception offsetting the initial cost. If we just once have to debug an obscure CMake issue or whatever else could go wrong with the generation, we could have just updated the file. I just don't see how this would ever be worth it both in terms of actual work needed to be done over time and in terms of complexity of the code-base we have to maintain.

**Comment by Flamefire** [comment]
That's why I wanted a meeting about this. I'm convinced the advantages outweigh the downsides. It'd be easier to discuss in person but let me still comment on that:

> This includes the additional CMake code that needs to be maintained

It's about 30 lines of pretty trivial code: Set variables, find the program, glob for *.tmpl and add a generator call for each of those. It is unlikely that this needs ever to be changed after the initial PR and even then it is pretty compact and easy to understand

> the template source files

The `reference` trait template file replaces 200 LOC with less then 50. So there is a considerate reduction of the code to be maintained, especially as that reduction can be assumed to be applicable to more parts than these 2 traits.

> als the project now requires clang-format to be built.

There are multiple options if you don't want that:

- Optional clang-format (would still work although the generated files won't be pretty, but the library still builds and is usable)
- Store the generated files in the repo (similar to otf2), so clang-format is a development requirement only
- Publish the generated files along the source with the github releases API. Can be done automatically for every commit to master, so people have an option if they cannot use clang-format.
- Require clang-format. As shown on travis this is very easy to satisfy on a normal linux or osx environment. Although I do realize, that it might not be as easy for clusters where a module would need to be installed or have the user download clang-format manually.

> I don't see how generating the small subset of regular code that requires no exception offsetting the initial cost

The current code has only empirical evidence of correctness, at best. [This whole file](https://github.com/tud-zih-energy/otf2xx/blob/c63a2648abd82152b14a59b3ec771b423610c460/include/otf2xx/traits/reference.hpp) is a sin: Magic values (the types) all over the place. There is no indication, that those copy&paste generated types are correct at all. Those, at least, have to be manually replaced by the OTF2 types, and even then one might make mistakes. [One example](https://github.com/tud-zih-energy/otf2xx/blob/c63a2648abd82152b14a59b3ec771b423610c460/include/otf2xx/reference.hpp#L123). This is wrong for all signed types and for the unsigned ones it relies on a defacto standard only. Besides that the template of the function is overly complex or wrong as it has no effect.   
The generated code is more compact and relies on the same mechanism used for the official interface. It can thus be assumed to be correct without using [400 lines of code just to test enum values](https://github.com/tud-zih-energy/otf2xx/blob/c63a2648abd82152b14a59b3ec771b423610c460/tests/enums_test.cpp) (Side note: 2 issues with that: The tests should all be `static_assert`s and it **must**  be build with the main code, or breaking it would never be noticed when used as a submodule, which e.g. you do)

>  If we just once have to debug an obscure CMake issue

As mentioned the CMake code is trivial. Unless you use paths with semicolons this is extremely unlikely.

> or whatever else could go wrong with the generation, we could have just updated the file

Most of the generation can be copied from the already in-use and tested generation of the OTF2-C-Files. And if anything major changes during an upgrade of OTF2, the code will stop compiling instead of silently being wrong. And if "something goes wrong" it is unlikely that it affects only a single point but instead it might require updates to multiple, related structs. This is much easier in the generated ones as one change fixes it for all instances instead of resorting to error-prone copy&paste.

>  I just don't see how this would ever be worth it both in terms of actual work needed to be done over time and in terms of complexity of the code-base we have to maintain.

Yes it obviously is work and it may or may not save work in the long run (IMO it will, because e.g. currently a lot would need to be replaced manually in tedious work  which could introduce random bugs. Generation would introduce only obvious bugs like missing structs)   
But one major point is, that the generated code is way more trustworthy than the C&P stuff that the code consists of ATM.

**Comment by bmario** [comment]
Superceeded be #33

**Pull request closed without merging**